[#2] Repository

jwonyLee opened the pull request:
## 개요

`PostEntity`를 다루는 데 사용하는 `Repository` 클래스 구현

## 작업 사항

- `PostEntity`와 대응하는 모델 객체 구현
- `PostRepository` 구현
  - `create(_:completion:)` 메서드 구현
  - `update(_:completion:)` 메서드 구현
  - `get(withIdentifier:completion:)` 메서드 구현
  - `fetchAll(completion:)` 메서드 구현
  - `deleteAll(completion:)` 메서드 구현
- `CategoryRepository` 구현
  - `fetchedResultsController` 프로퍼티
  - `create(_:completion:)` 메서드 구현
  - `find(with:completion:)` 메서드 구현
  - `fetchAll(completion:)` 메서드 구현
  - `deleteAll(completion:)` 메서드 구현

`delete` 메서드를 구현하지 못했는데, 이후에 휴지통 기능 구현하면서 같이 구현하겠습니다. 😂

### Linked Issue

close #2

jwonyLee committed:
[#2] test: Post를 하나 만들 수 있는 테스트 작성

jwonyLee committed:
[#2] feat: Post 구조체 생성

jwonyLee committed:
[#2] feat: create new Post

jwonyLee committed:
[#2] feat: Merge branch 'upstream/develop'

jwonyLee committed:
[#2] feat: save Post in CoreData

jwonyLee committed:
[#2] feat: delete All & fetch all Post in CoreData

jwonyLee committed:
[#2] feat: get post with uuid in CoreData

jwonyLee committed:
[#2] feat: update post in Core Data

jwonyLee committed:
[#2] chore: Merge branch from ThingLog/develop

jwonyLee committed:
[#2] fix: 피드백 반영
- Attachment.ImageData.blob -renaming-> Attachment.ImageData.originalImage
- fetchAll : mainContext.perform 으로 변경
- RepositoryTests 변경
- get : mainContext.perform 으로 변경
- update : get 메서드 사용하는게 변경
- 주석 추가
- Result<_, Error> -> Result<_, PostRepositoryError>로 변경

jwonyLee committed:
[#2] fix: 2차 피드백 반영

jwonyLee committed:
[#2] chore: PostRepositoryError 파일을 별도로 분리

jwonyLee committed:
[#2] feat: fetch All Category

jwonyLee committed:
[#2] feat: Create Category Entity

jwonyLee committed:
[#2] feat: CategoryRepository 별도 분리

jwonyLee committed:
[#2] feat: Category 관련 로직 추가
- find(with:completion:)
- deleteAll(completion:)

jwonyLee committed:
Merge branch 'develop' into feature/2-repository

gustn3965 commented on ThingLog/Repository/CategoryRepositoryError.swift:
```diff
@@ -0,0 +1,27 @@
+
+import Foundation
+
+enum CategoryRepositoryError: Error {
```
👀 `PostRepositoryError` 와 동일한 에러가 작성됐는데, 굳이 `Category**Error`로 따로 에러를 만들 필요가 있을까욥 ?  `Cateogory`도 하나의 `NSManagedObject`로써 의미는 같다라고 생각돼서요~

gustn3965 commented on ThingLogTests/ThingLogRepositoryTests.swift:
```diff
@@ -0,0 +1,356 @@
+        }
+    }
+
+    func test_Category를_하나_추가할_수_있다() {
```
⚠️ Test에 Test코드가 굉장히 많아지기도 하고, `Category`관련된 Test는 따로 `CategoryRepositoryTest.swift` 로 만들어서 따로 작성하는게 좋을 듯 해용

gustn3965 commented on ThingLog/Repository/CategoryRepository.swift:
```diff
@@ -0,0 +1,112 @@
+                if let find: CategoryEntity = try fetchRequest.execute().first {
+                    completion(true, find)
+                }
+                completion(false, nil)
```
⚠️ 성공하여도 `completion(false, nil)`이 호출될 거에요.

gustn3965 commented on ThingLog/SupportingFiles/ThingLog.xcdatamodeld/ThingLog.xcdatamodel/contents:
```diff
@@ -14,43 +14,38 @@
         <attribute name="title" attributeType="String"/>
         <relationship name="attachments" optional="YES" toMany="YES" deletionRule="Cascade" destinationEntity="Attachment" inverseName="post" inverseEntity="Attachment"/>
-        <relationship name="categories" toMany="YES" deletionRule="Cascade" destinationEntity="Category" inverseName="post" inverseEntity="Category"/>
+        <relationship name="categories" toMany="YES" deletionRule="Cascade" destinationEntity="Category" inverseName="posts" inverseEntity="Category"/>
```
👀 `cascade`이면 `PostEntity`를 삭제할 경우 `CategoryEntity`도 같이 삭제되지 않나용? `CategoryEntity `는 유니크해야하지 않나용??

jwonyLee replied:
`Cascade`에서 `Nullify`로 변경했습니다!

gustn3965 commented on ThingLog/Repository/CategoryRepository.swift:
```diff
@@ -0,0 +1,112 @@
+        self.fetchedResultsControllerDelegate = fetchedResultsControllerDelegate
+    }
+
+    lazy var fetchedResultsController: NSFetchedResultsController<CategoryEntity> = {
```
👀 `fetchedResultsController`가 필요한 이유를 알 수 있을까용?  또한 `delegate`가 필요한 이유도요~

jwonyLee replied:
화면 설계서에서 카테고리 페이지를 별도로 두고 있는데요, 이 때 목록에서 바로 출력해버릴까 하는 생각으로 만들었습니당. `delegate`는 태그가 추가 되었을 때 테이블뷰를 reload 하기 위해서 추가 했어욥

gustn3965 commented on ThingLogTests/ThingLogRepositoryTests.swift:
```diff
@@ -199,6 +207,7 @@ class ThingLogRepositoryTests: XCTestCase {
 
     func test_Post를_수정할_수_있다() {
```
👀 `Post`를 카테고리를 포함해서 작성하고,` Post`의 카테고리를 변경하거나 삭제하는 테스트 코드와,  작성완료한 `Post `의 카테고리를 변경해도 해당 카테고리는 남아있는지 테스트 코드도 필요할 것 같아요.

gustn3965 commented on ThingLogTests/ThingLogRepositoryTests.swift:
```diff
@@ -0,0 +1,356 @@
+                    }
+                }
+            }
+            exp.fulfill()
```
⚠️ 이부분은 디스코드로 따로 말씀드렸습니당

gustn3965 commented on ThingLog/Repository/CategoryRepository.swift:
```diff
@@ -0,0 +1,112 @@
+        let fetchRequest: NSFetchRequest<CategoryEntity> = CategoryEntity.fetchRequest()
+        fetchRequest.predicate = NSPredicate(format: "title == %@", categoryName)
+
+        coreDataStack.mainContext.performAndWait {
```
👀🥲 `performAndWait` 를 사용한 이유를 `find`가 사용된 코드를 보고서 이해할 수 있었습니당.   평소와 다른 메서드가 사용될 경우에는 그에 대한 정보를 주석으로 제공해주시면 감사하겠습니다. ㅠ

gustn3965 commented on ThingLogTests/ThingLogRepositoryTests.swift:
```diff
@@ -229,23 +238,109 @@ class ThingLogRepositoryTests: XCTestCase {
+
+    func test_모든_Category를_가져올_수_있다() {
+        // given: 필요한 모든 값 설정
+        deleteAllEntity()
```
⚠️ `deleteAllEntity`가 다 `backgroundTask`로 실행되고 있기 때문에,  다 삭제가 되기도 전에  다음 코드로 실행될 수 있습니당

gustn3965 commented on ThingLogTests/ThingLogRepositoryTests.swift:
```diff
@@ -229,23 +238,109 @@ class ThingLogRepositoryTests: XCTestCase {
+            }
+            exp.fulfill()
+        }
+
```
⚠️ 이부분도  디코로 설명드린것과 같습니다. `forEach` -`create`

gustn3965 commented on ThingLogTests/ThingLogRepositoryTests.swift:
```diff
@@ -229,23 +238,109 @@ class ThingLogRepositoryTests: XCTestCase {
+        }
+    }
+
+    func test_모든_Category를_가져올_수_있다() {
```
👀 `test_모든_Category를_가져올_수_있다` 보다는 좀 더 테스트 상황과 유사하게 정보를 더 제공 해주시면 좋을 것 같아용   `test_3개__Category를_만들어서3개_Category를_가져올_수_있다`

gustn3965 commented on ThingLog/Entity/PostEntity+.swift:
```diff
@@ -17,6 +17,12 @@ extension PostEntity {
         self.rating?.scoreType = post.rating.score
-        self.categories = NSSet(array: [post.categories.map { $0.toEntity(in: context) }])
+        if let categories = self.categories {
+            self.removeFromCategories(categories)
```
👀 이부분이 걱정됩니다. `cascade`로 되어있어서 ( 이부분도 따로 코멘트 달았어요 ) 다른 `Post`와 연결된 카테고리마저 다 삭제될 것 같아서용.

gustn3965 reviewed (commented):
고생하셨습니다!  최대한 꼼꼼하게 하느라 시간이 좀 걸렸네용. 리뷰 확인해주세욥 ~

jwonyLee committed:
[#2] fix: 피드백 반영
- RepositoryError 로 통일
- CategoryRepositoryTest 별도로 분리
- deleteAllEntity() 를 timeout()으로 wrapping
- 테스트 코드 fulfill() 위치 변경
- fatalError()를 XCTFail()로 변경

jwonyLee committed:
Merge branch 'feature/2-repository' of https://github.com/jwonylee/thinglog into feature/2-repository

jwonyLee committed:
[#2] test: CategoryRepository 관련 테스트 코드 추가

gustn3965 commented on ThingLogTests/ThingLogPostRepositoryTests.swift:
```diff
@@ -0,0 +1,391 @@
+        }
+    }
+
+    func test_Post를_3개_만들고_3개의_Post를_가져올_수_있다() {
```
👀  `test_Post를_3개_만들_수_있다` 와 차이가 없어보이는데 하나만 하셔도 괜찮을 것 같아용~

jwonyLee committed:
[#2] fix: repository context 변경

gustn3965 reviewed (approved):
고생하셨습니당 ! 
애초에 구조를 잘 짜신 것 같아서 불필요한 코드의 수정은 거의 없고 컨텍스트는 간단하게 수정하면 되는 코드만 거의 있네용 

디코로도 말씀드렸듯이 다음에는 테스트코드도 더 다양하게 더 세세하게 작성해주시면 좋을 것 같습니당 ~ ! 👍

jwonyLee committed:
Merge branch 'develop' into feature/2-repository
